include/swift/SwiftRemoteMirror/SwiftRemoteMirror.h

```diff
@@ -80,6 +80,12 @@ void
 swift_reflection_addReflectionInfo(SwiftReflectionContextRef ContextRef,
                                    swift_reflection_info_t Info);
 
+/// Add reflection sections that may not be contiguous from a loaded Swift image.
+SWIFT_REMOTE_MIRROR_LINKAGE
+void swift_reflection_addNonContiguousReflectionInfo(
```

Comment:
Is there a conflicting interface for `swift_reflection_addReflectionInfo`?  I'm not sure I see the value in the `NonContiguous` here.  Or was there some other reason that you opted to add that?

Reply:
The difference is that this receives a `swift_non_contiguous_reflection_info_t` instead of a `swift_reflection_info_t`. I needed to change the data structure because `swift_reflection_info_t` expects the offset of the local and remote sections to be the same (because it assumed the sections were contiguous, which isn't true).

`swift_reflection_addReflectionInfo` already exists, and @vedantk asked me to not change it since it's part of the API.

Reply:
I believe `swift_reflection_info_t` isn't sufficiently general to describe a non-contiguous section layout.

Reply:
Under what circumstances are you seeing a non-contiguous metadata section?

Reply:
@jckarter in my tests on Linux, they always appear to be non-contiguous. Maybe it's OS related?

Reply:
Ah, I see what you mean, the subsections are contiguous, but the mapping from remote to local addresses between them may not be 1:1. That's reasonable. As the comments note, `swift_reflection_addReflectionInfo` is deprecated and only exists for legacy ABI reasons on Apple platforms, because it's unable to accurately express this reality. We could probably remove it entirely on non-Apple platforms. It would be great if we could name this new API in a way that makes it more apparent it should be the preferred API for new code using Remote Mirror that wants to manually register images.

Reply:
I might be wrong, but from my tests I believe the subsections really weren't contiguous. I'll re-check this to see if I'm correct.

Reply:
@jckarter here is what `swift-reflection-test.c` logs in my machine (I just added the End field, so it's easier to visualize):

```
makeRemoteSection: Making remote section with Start = 0x7f22a7aee7b0 End = 0x7f22a7af285c and  Size = 16556
makeRemoteSection: Making remote section with Start = 0x7f22a7ae99a8 End = 0x7f22a7aee7b0 and  Size = 19976
makeRemoteSection: Making remote section with Start = 0x7f22a7a13f1c End = 0x7f22a7a14a98 and  Size = 2940
makeRemoteSection: Making remote section with Start = 0x7f22a7ae8f40 End = 0x7f22a7ae99a8 and  Size = 2664
makeRemoteSection: Making remote section with Start = 0x7f22a7a0d864 End = 0x7f22a7a12e43 and  Size = 21983
makeRemoteSection: Making remote section with Start = 0x7f22a7a12e50 End = 0x7f22a7a13f19 and  Size = 4297

makeRemoteSection: Making remote section with Start = 0x7f22a7334244 End = 0x7f22a7334244 and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a7334244 End = 0x7f22a7334244 and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a7334244 End = 0x7f22a7334244 and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a7334244 End = 0x7f22a7334244 and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a7349180 End = 0x7f22a734927c and  Size = 252
makeRemoteSection: Making remote section with Start = 0x7f22a7334244 End = 0x7f22a7334244 and  Size = 0

makeRemoteSection: Making remote section with Start = 0x7f22a74f640c End = 0x7f22a74f6a40 and  Size = 1588
makeRemoteSection: Making remote section with Start = 0x7f22a74f63f4 End = 0x7f22a74f640c and  Size = 24
makeRemoteSection: Making remote section with Start = 0x7f22a74f3b1c End = 0x7f22a74f3b1c and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a74f3b1c End = 0x7f22a74f3b1c and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a74f63d0 End = 0x7f22a74f63f3 and  Size = 35
makeRemoteSection: Making remote section with Start = 0x7f22a74f36a9 End = 0x7f22a74f3b05 and  Size = 1116

makeRemoteSection: Making remote section with Start = 0x7f22a7b25344 End = 0x7f22a7b254a4 and  Size = 352
makeRemoteSection: Making remote section with Start = 0x7f22a7b2530c End = 0x7f22a7b25344 and  Size = 56
makeRemoteSection: Making remote section with Start = 0x7f22a7b21cce End = 0x7f22a7b21cce and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a7b21cce End = 0x7f22a7b21cce and  Size = 0
makeRemoteSection: Making remote section with Start = 0x7f22a7b25228 End = 0x7f22a7b25309 and  Size = 225
makeRemoteSection: Making remote section with Start = 0x7f22a7b21c00 End = 0x7f22a7b21cbe and  Size = 190

makeRemoteSection: Making remote section with Start = 0x560e295150dc End = 0x560e295150f8 and  Size = 28
makeRemoteSection: Making remote section with Start = 0x560e295131ee End = 0x560e295131ee and  Size = 0
makeRemoteSection: Making remote section with Start = 0x560e295131ee End = 0x560e295131ee and  Size = 0
makeRemoteSection: Making remote section with Start = 0x560e295131ee End = 0x560e295131ee and  Size = 0
makeRemoteSection: Making remote section with Start = 0x560e295150d0 End = 0x560e295150dc and  Size = 12
makeRemoteSection: Making remote section with Start = 0x560e295131ec End = 0x560e295131ee and  Size = 2
```
I think the second block is the easiest to look at, since all sections, except 1, have size 0. The old algorithm would mistakenly assume that the sections would go from address 0x7f22a7334244 up to 0x7f22a734927c, which is 15038 bytes long.

Reply:
Sorry, by "subsection" I mean one of the individual areas referenced in the `reflection_info` struct, like `field`, `associated_types`, and so on. Anyway, the `swift_reflection_info` and `swift_reflection_addReflectionInfo` interfaces are obsolete, and I'm saying it would be good to name these new ones `swift_reflection_addReflectionInfo2` or something like that to make it clear it should supersede the old interface.